Fix multithreading in Geant4 and HitManager

sethrj opened the pull request:
This adds thread safety to the HitCollector and fixes #613 .

After this fix, here's a comparison of 18 GeV $\pi^+$ in the ATLAS TileCal model:
![18gev-piplus](https://user-images.githubusercontent.com/741229/226917236-dc3886c3-e6a6-4a2f-aa92-339dbacbd508.png)

sethrj committed:
Create thread-local hit processors

sethrj committed:
Add debug log messages to sensitive detectors

sethrj committed:
Make StepCollector thread-safe

sethrj committed:
Extend InitializedValue to do finalization

sethrj committed:
Use InitializedValue to correctly finalize local data

whokion commented on test/corecel/cont/InitializedValue.test.cc:
```diff
@@ -0,0 +1,129 @@
+    // constructing into data space that's been set to a different value
+    alignas(int) Byte buf[sizeof(int)];
+    std::fill(std::begin(buf), std::end(buf), Byte(-1));
+    InitValueInt* ival = new (buf) InitValueInt{};
```
error: placement new constructing an object of type ‘InitValueInt’ {aka ‘celeritas::InitializedValue<int>’} and size ‘8’ in a region of type ‘celeritas::Byte [4]’ and size ‘4’ [-Werror=placement-new=]
   28 |     InitValueInt* ival = new (buf) InitValueInt{};

sethrj commented:
Sorry, I shouldn't have requested the review yet because this is still in draft and includes changes from #693 that don't need review here.

whokion commented:
No problem. Let's me know when it is ready to review.

whokion commented on src/accel/LocalTransporter.cc:
```diff
@@ -50,7 +51,9 @@ inline Real3 convert_from_geant(G4ThreeVector const& vec, double units)
-    : auto_flush_(options.max_num_tracks), max_steps_(options.max_steps)
+    : auto_flush_(options.max_num_tracks)
+    , max_steps_(options.max_steps)
+    , hit_manager_{params.hit_manager()}
```
It may be a design choice, but why is `HitManager` part of `SharedParams`? Should it be a just thread local object?

sethrj replied:
`HitManager` is shared across threads (it's called by the shared "action" inside the stepping loop), but it also needs to be individually accessed by each thread because of how Geant4 deallocates memory... so making it accessible from the local transporter was the best way I could think of doing it.

whokion replied:
Isn't the stepping loop (so the shared "action") belong to each transporter which returns a hit collection at the end of kernel?  So, why can transporter (which is thread local) just own a private Hit Manager (so, thread local) which in turn has a private HitProcess?  So, the question is why HitManager should be shared by all threads?

sethrj replied:
Even though the "loop" is stored by each stepper, the actions are all shared in coreparams.

whokion commented on src/accel/LocalTransporter.hh:
```diff
@@ -78,6 +91,9 @@ class LocalTransporter
     size_type auto_flush_{};
     size_type max_steps_{};
+
+    // Shared pointer across threads, "finalize" called when clearing
```
Isn't hit_manager_ thread local (i.e., why is it shared across threads)?

sethrj replied:
No, it's owned by the `StepCollector` which is global (it ties into the stepping callbacks). The hit *processors* are thread-local though.

whokion commented on src/accel/detail/HitManager.cc:
```diff
@@ -168,11 +183,43 @@ void HitManager::execute(StateHostRef const& data)
+/*!
+ * Ensure the local hit processor exists, and return it.
+ */
+HitProcessor& HitManager::get_local_hit_processor()
```
If HitManager is thread local, HitProcess can be private to HitManager (related to the early comment) instead of vector<HitProcess*>.  Why is this the better approach?

sethrj replied:
All the major event loop components are "global". Only the incoming state objects are "thread-local" [^1], and normally their stream IDs should be used to access per-stream state data that (for now) has to live in the shared object. Here, instead of using G4Threading, I could (and probably should) add the `StreamId` to the `StepState` so we can access it directly...

[^1]: I put global and thread-local in quotes above because nothing in Celeritas inherently requires streams and threads to match. Streams could be shared within a thread pool, or you could have multiple streams per CPU thread, etc. Similarly you could have two entirely different event loops running simultaneously within the same application.

whokion commented on src/accel/detail/HitProcessor.cc:
```diff
@@ -168,14 +168,57 @@ HitProcessor::HitProcessor(VecLV detector_volumes,
+    for (auto i : range(detectors_.size()))
+    {
+        CELER_ASSERT((*detector_volumes_)[i]);
+        detectors_[i] = (*detector_volumes_)[i]->GetSensitiveDetector();
```
Is this assuming that all sensitive volumes are sensitive (and unique)?

sethrj replied:
This input is based the detector list constructed by the HitManager, which found all the logical volumes that have attached sensitive detectors. This just acts as the translation from logical volume to attached (thread-local) sensitive detector.

sethrj commented:
@whokion This is ready to review whenever you have a moment. Specifically could you make sure that the way I'm using `G4Threading` is compatible with what you know about tasking in CMSSW? Given my cursory look (CMSSW changes the thread-local ID in geant to values in `[0, N)`) it should be compatible. Thanks!

whokion commented on src/accel/detail/HitProcessor.hh:
```diff
@@ -32,48 +34,53 @@ namespace detail
- * action manager.
+ * generating hit objects. It \b must be thread-local because the sensitive
+ * detectors it stores are thread-local, and additionally Geant4 makes
+ * assumptions about object allocations that cause crashes if the HitProcessor
```
A worker has its own event loops.  So, this is not an assumption  as the Geant4 MT is an event-level multithreading, but a consequence.

sethrj replied:
The "assumption" is that objects like G4Navigator, G4Step, and G4Track can *only* be allocated by a worker thread, and that they're deallocated by the same worker thread. Those objects aren't necessarily associated with a specific event (in the G4 fast simulation some of those are reused across multiple events), nor are they inherently associated with a specific thread (except by the invisible implementation of the thread allocator).

whokion commented on src/accel/detail/HitProcessor.cc:
```diff
@@ -230,26 +273,21 @@ void HitProcessor::operator()(DetectorStepOutput const& out) const
         {
-            CELER_ASSERT(out.detector[i] < detector_volumes_.size());
+            CELER_ASSERT(out.detector[i] < detectors_.size());
             bool success = this->update_touchable(
```
This seems very expensive. Do we really need this check?

sethrj replied:
The touchable update is necessary for detectors like tilecal and hgcal that use the detailed navigation state to determine a subdetector identifier. Since their sensitive detectors require the navigator to be initialized and correct, we don't really have a choice if we're calling back to those routines. My hope is that for our initial implementation this won't completely kill the performance, and doing it this way will give us a "baseline" performance number to justify the effort of implementing the actual detector logic on GPU.

whokion replied:
I still do not understand why HitProcess is responsible for updating the navigation state (touchable) as a hit is processed at the end of stepping (only once per each step) - isn't the subdetector a physical/logical volume (so can be a tracking volume  with a boundary)?  Isn't it an independent readout channel?  I guess that there may be a confusion between what simulation information needs to be known (for MC "study") and what need to be recorded as hit (to be used for digitization).

sethrj replied:
I might be misunderstanding what the confusion is too, so let me try to restate from a high level what's going on.
- At the end of an event on a local thread, the `LocalTransporter` steps through all the offloaded EM tracks until they die.
- Each step calls the (shared) explicit actions using the shared "params" state and the (thread-local) track states.
- The `celeritas::StepCollector` gathers the position and volume ID at the pre-step and the energy deposition at post-step (though the exact selection of outputs can be changed by the user).
- At post-step, the (shared) HitManager takes the thread-local gathered step data, and calls the thread-local HitProcessor.
- The HitProcessor loops through all the hits, translates them into a thread-local `G4Step` objects, uses the VecGeom logical volume ID to look up the `G4VSensitiveDetector`, and calls `sd->Hit(step)`. If the SD needs to query `hit->GetPreStepPoint()->GetTouchableHandle()` then before calling `Hit` we use a thread-local (i.e., owned by HitProcessor) navigator to update the thread-local touchable for the thread-local step.

When I said "subdetector" I meant "readout channel": I guess HGCal is the subdetector, and the codeneeds the navigation state and volume information to figure out what channel it's in.

whokion commented:
Can you add a code snippet how to access the collected hit information and retrieve some information (for an example, the total energy deposition or the size of hit collection from at the end of EventAction (or after each Flush?) into the demo-geant-integration?  For CMS integration, we need to access the hit (StepInfo) collection from GPU at least in the event level to merge hits from CPU, and send them to the next stage of the simulation pipeline (i.e., digitization) or write them out to the disk in an event unit.

sethrj commented:
@whokion For our initial CMS implementation, we're *not* going to merge the hits on the GPU; we're using the `HitProcessor` to call back to the existing sensitive detectors. That way we don't have to reimplement the complex hit processing in CMSSW, and because we can call to the other SDs the same way, we can track EM particles everywhere rather than in a particular region. The existing `demo-geant-integration` does this too: the sensitive detector is getting hits from the GPU via the HitProcessor.

I *am* working on implementing a GPU-based calorimeter which will actually do the accumulation on GPU—is this what you're asking for?

whokion commented:
> I _am_ working on implementing a GPU-based calorimeter which will actually do the accumulation on GPU—is this what you're asking for?

No, the current implementation/workflow is good enough for a demonstration/integration.  What I am asking is to pipeline the output from GPU (to create a collection of equivalent G4Steps which will be merged through CPU sensitive detectors into the final hit data) into the demo-geant-integration workflow once the transporter complete stepping for the set of EM tracks on GPU (i.e. after Flush is called) - at least, an example how to access the output hit information from GPU in EventAction or TrackingAction will be good enough.

sethrj commented:
But this is already happening automatically through the HitManager during the stepping loop. The current demo app will have Steps sent back from GPU.

whokion commented:
Then, I definitely missed something.  So, are we calling `demo_geant::SensitiveDetector::ProcessHits(G4Step* step, G4TouchableHistory*)` with converted Steps from GPU?  In `demo-geant-intergration`, where is the place receiving output of `HitManager::selection()` and converting `StepSelection::points` to G4Steps, and processing them with SensitiveDetector::ProcessHits inside a loop over the number of hits from GPU?

sethrj commented:
We are. There are a lot of moving pieces, so I'll draw up a diagram of who's constructing what and how the tracks and hits move between CPU and GPU 😄

sethrj commented:
@whokion Here's my attempt at a UML diagram, with background colors indicating the different libraries/layers of code, and a few classes that operate on device in green.

![hit-processor-uml](https://user-images.githubusercontent.com/741229/227957440-738f57a6-9c8a-4357-9658-3d32536ac95d.png)

The user tracking action (local) sends hits via G4Track to the `LocalTransporter` and the user event action (local) flushes it, which initiates the `ActionSequence` (stepping loop), managed by the Stepper (local). The `ActionSequence` is just a topologically sorted graph of `ExplicitActionInterface` (shared), each of which usually represents a single kernel.

During the stepping loop (which uses the stream-local `CoreStateData` and shared `CoreParamsData`) the `StepGatherAction` (shared) fills up a `StepStateData` (local) with track properties like position, logical volume, etc. At the end of the step, all of the `StepInterface` classes (shared) originally registered with the `StepCollector` (shared) are called with the `StepStateData`. The `HitManager` (shared) is a `StepInterface` class that has stream-local `HitProcessor` instances. The `HitProcessor` copies the GPU hits to CPU `DetectorStepOutput` (local) hits, converts each hit into a `G4Step`, and calls the associated (local) `G4VSensitiveDetector`.

amandalund reviewed (approved):
Nice! This looks good to me @sethrj.

whokion commented:
> The user tracking action (local) sends hits via G4Track to the `LocalTransporter` and the user event action (local) flushes it, which initiates the `ActionSequence` (stepping loop), managed by the Stepper (local). 

Technically, the tracking action actually flushes (through LocalTransporter::Push(G4Track), then ::Flush() and the event action flushes the remaining of tracks at the end of event.

Nevertheless, I am still not fully convinced why HitManager is globally shared as the Stepper/LocalTransporter are local even though HitManager behaviors like something similar to Geant4 Split class mechanism, but find nothing wrong either - may be a design choice.  Maybe we may learn more how to directly use the HitManager/HitProcessor for the CMS integration as necessary.

Anyway, the diagram helps a lot and enlightens inter-connectivity and workflow with specific ownership/relations.  Thanks for the nice work.

sethrj commented:
OK @whokion a minor update that I think improves even further:

![hit-processor-uml](https://user-images.githubusercontent.com/741229/228025417-579495b6-a629-4f8b-bc95-0a33eb04a5d9.png)

For clarity I switched StepCollector and StepStateData and fixed the ownership relationship between HitProcessor and DetectorsStepOutput. The dotted region are the classes that are shared across threads/tasks/streams because they operate only on the problem setup/parameter data rather than any state information.  The HitManager is-a StepInterface derived class that is managed by the (shared) step gather action, so it has to be shared. The HitProcessor operates on the state data so it has to be stream-local.

Due to the way Geant4 currently manages memory, we *have* to make sure that our streams correspond to geant4 threads. There's a hidden "references" link here between the LocalTransporter and the HitManager which is necessary to ensure that the hit processor's temporary data is deallocated by the thread-local `LocalTransporter`, made necessary by the weird Geant4 allocation, which is part of the weirdness in the design that you may have noticed.
